src/Features/Core/Portable/EditAndContinue/EditSession.cs

```diff
@@ -342,16 +341,62 @@ private static async Task PopulateChangedAndAddedDocumentsAsync(CommittedSolutio
                 changedOrAddedDocuments.Add(document);
             }
 
-            foreach (var documentId in changes.GetAddedDocuments())
+            foreach (var documentId in newProject.State.DocumentStates.GetAddedStateIds(oldProject.State.DocumentStates))
             {
-                var document = project.GetDocument(documentId)!;
+                var document = newProject.GetDocument(documentId)!;
                 if (document.State.Attributes.DesignTimeOnly)
                 {
                     continue;
                 }
 
                 changedOrAddedDocuments.Add(document);
             }
+
+            // TODO: support document removal/rename (see https://github.com/dotnet/roslyn/issues/41144, https://github.com/dotnet/roslyn/issues/49013).
+
+            // Given the following assumptions:
+            // - source generators are deterministic,
+            // - source documents, metadata references and compilation options have not changed,
+            // - additional documents have not changed,
+            // - analyzer config documents have not changed,
+            // the outputs of source generators will not change.
+            // 
+            // Currently it's not possible to change compilation options (Project System is readonly during debugging).
```

Comment:
I wonder what happens if we use our code fixes that change compilation options. Not sure how much I want to try though!

Reply:
Probably nothing good if the options affect emit. We'll need to handle that some day, but not today :)